python/demo/demo_navier_stokes.py

```diff
@@ -0,0 +1,419 @@
+
+
+def f_expr(x):
+    """Expression for the applied force"""
+    return np.vstack((np.zeros_like(x[0]),
+                      np.zeros_like(x[0])))
+
+
+def boundary_marker(x):
+    return np.isclose(x[0], 0.0) | np.isclose(x[0], 1.0) | \
+        np.isclose(x[1], 0.0) | np.isclose(x[1], 1.0)
+
+
+# We define some simulation parameters
+
+n = 16
+num_time_steps = 25
+t_end = 10
+R_e = 25  # Reynolds Number
+k = 1  # Polynomial degree
+
+# Next, we create a mesh and the required functions spaces over
+# it. Since the velocity uses an $H(\textnormal{div}$-conforming function
+# space, we also create a vector valued discontinuous Lagrange space
+# to interpolate into for artifact free visualisation.
+
+msh = mesh.create_unit_square(MPI.COMM_WORLD, n, n)
+# Function space for the velocity
+V = fem.FunctionSpace(msh, ("Raviart-Thomas", k + 1))
+# Function space for the pressure
+Q = fem.FunctionSpace(msh, ("Discontinuous Lagrange", k))
+# Funcion space for visualising the velocity field
+W = fem.VectorFunctionSpace(msh, ("Discontinuous Lagrange", k + 1))
+
+# Define trial and test functions
+
+u, v = TrialFunction(V), TestFunction(V)
+p, q = TrialFunction(Q), TestFunction(Q)
+
+delta_t = fem.Constant(msh, PETSc.ScalarType(t_end / num_time_steps))
+alpha = fem.Constant(msh, PETSc.ScalarType(6.0 * k**2))
+R_e_const = fem.Constant(msh, PETSc.ScalarType(R_e))
+
+h = CellDiameter(msh)
+n = FacetNormal(msh)
+
+
+def jump(phi, n):
+    return outer(phi("+"), n("+")) + outer(phi("-"), n("-"))
+
+
+# We solve the Stokes problem for the initial condition, so the convective
+# terms are omitted for now
+
+a_00 = 1 / R_e_const * (inner(grad(u), grad(v)) * dx
+                        - inner(avg(grad(u)), jump(v, n)) * dS
+                        - inner(jump(u, n), avg(grad(v))) * dS
+                        + alpha / avg(h) * inner(jump(u, n), jump(v, n)) * dS
+                        - inner(grad(u), outer(v, n)) * ds
+                        - inner(outer(u, n), grad(v)) * ds
+                        + alpha / h * inner(outer(u, n), outer(v, n)) * ds)
+a_01 = - inner(p, div(v)) * dx
+a_10 = - inner(div(u), q) * dx
+a_11 = fem.Constant(msh, PETSc.ScalarType(0.0)) * inner(p, q) * dx
+
+a = fem.form([[a_00, a_01],
+              [a_10, a_11]])
+
+f = fem.Function(W)
+u_D = fem.Function(V)
+u_D.interpolate(u_e_expr)
+L_0 = inner(f, v) * dx + \
+    1 / R_e_const * (- inner(outer(u_D, n), grad(v)) * ds
+                     + alpha / h * inner(outer(u_D, n), outer(v, n)) * ds)
+L_1 = inner(fem.Constant(msh, PETSc.ScalarType(0.0)), q) * dx
+
+L = fem.form([L_0,
+              L_1])
+
+# Boundary conditions
+
+boundary_facets = mesh.locate_entities_boundary(
+    msh, msh.topology.dim - 1, boundary_marker)
+boundary_vel_dofs = fem.locate_dofs_topological(
+    V, msh.topology.dim - 1, boundary_facets)
+bc_u = fem.dirichletbc(u_D, boundary_vel_dofs)
+
+# The pressure is only determined up to a constant, so pin a single degree
```

Comment:
Is this compatible with the constraint stated above that `\int_\Omega p \mathrm{d}\vec{x} = 0`?

Reply:
No, it's just a simple way to fix the pressure in the discrete problem, but wouldn't be mathematically valid in the continuous problem. I could subtract the average off the pressure, as I do when computing the error norm. Alternatively, I could change the solver setting and remove the nullspace without pinning a pressure dof.